### ISSUE TITLE
Separate "possibly unbound" and "certainly unbound" diagnostics

### ISSUE THREAD

**Issue by daquexian** [microsoft/pyright#6896]
**Is your feature request related to a problem? Please describe.**
There are many issues about the false positive "possibly unbound" diagnostics, like https://github.com/microsoft/pyright/issues/3041 and https://github.com/microsoft/pyright/issues/4087. I know it is designed but these false positive cases greatly bother me. Meanwhile, I don't want to suppress the "certainly" unbound diagnostics, which always mean a bug.

**Describe the solution you’d like**

Separate "possibly unbound" and "certainly unbound" diagnostics, so that I can write a pyrightconfig.json like

```json
{
  "reportCertainlyUnboundVariable": "error",
  "reportPossiblyUnboundVariable": "information",
}
```

**Comment by erictraut** [comment]
I understand your use case, but making the change you're suggest would be a breaking change for existing pyright users who have configured their projects with the current `reportUnboundVariable`. That raises the bar for making such a change.

There are simple workarounds for avoiding possibly-unbound variables. Are these workarounds not feasible in your code?

I'll think about this a bit more, but my inclination is to decline with the option of revisiting it in the future if we receive a stronger signal that this is a desirable change.

**Comment by daquexian** [comment]
> There are simple workarounds for avoiding possibly-unbound variables. Are these workarounds not feasible in your code?

I'm working with a forked, large codebase. The false positive diagnostics from the existing part greatly hide the real diagnostics from our own part. I prefer not to modify the existing part because it brings potential conflicts with upstream.

**Comment by erictraut** [comment]
I'm going to reject this idea for now. As I mentioned, this change would create some pain for current pyright users, and I don't think the justification is sufficiently compelling to overcome these downsides.

I'm open to reconsidering this in the future based on additional feedback from pyright users.

**Comment by erictraut** [comment]
I'm doing a major pass over all of the diagnostic rules, especially `reportGeneralTypeIssues`, and creating finer-grained rules. This is a good time to make the change requested in this ticket. I've added a new `reportPossiblyUnboundVariable` rule, and this will be included in the next release of pyright.
